README.md
# clojure_katas

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-05-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-05-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure_katas/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure_katas/compare/0.1.0...0.1.1


test/clojure_katas/core_test.clj
(ns clojure-katas.core-test
  (:require
   [clojure-katas.core :refer :all]
   [midje.sweet :refer [facts fact provided => roughly anything]]))

(facts "test fizz buzz"
       (fact "zero"
             (fizzBuzz 0) => 0)
       (fact "one"
             (fizzBuzz 1) => 1)
       (fact "fizz"
             (fizzBuzz 3) => "fizz")
       (fact "buzz"
             (fizzBuzz 5) => "buzz")
       (fact "fizz multiple"
             (fizzBuzz 6) => "fizz")
       (fact "buzz multiple"
             (fizzBuzz 10) => "buzz")
       (fact "fizzBuzz"
             (fizzBuzz 15) => "fizzbuzz"))

(facts "test factorial"
       (fact "1"
             (factorial 1) => 1)
       (fact "2"
             (factorial 2) => 2)
       (fact "3"
             (factorial 3) => 6)
       (fact "4"
             (factorial 4) => 24))

(facts "test stack"
       (fact "pop empty stack returns empty"
             (spop) => nil)
       (fact "pop with one element return element"))
             ;(push 0)
             ;(spop) => 0

(facts "test multiple 3 and 5"
       (fact "given 0"
             (multiples3And5 0) => 0)
       (fact "given 1"
             (multiples3And5 1) => 0)
       (fact "given 3"
             (multiples3And5 3) => 0)
       (fact "given 4"
             (multiples3And5 4) => 3)
       (fact "given 5"
             (multiples3And5 5) => 3)
       (fact "given 6"
             (multiples3And5 6) => 8)
       (fact "given 10 returns 23"
             (multiples3And5 10) => 23))

(facts "test equal sum of fibonacci numbers"
       (fact "given 0"
             (sumOfEvenFibonacciNumbers 0) => 0)
       (fact "given 1"
             (sumOfEvenFibonacciNumbers 1) => 0)
       (fact "given 2"
             (sumOfEvenFibonacciNumbers 2) => 0)
       ;; (fact "given 3"
       ;;       (sumOfEvenFibonacciNumbers 3) => 2)
       (fact "given 11"
             (sumOfEvenFibonacciNumbers 11) => 44))

(facts "test fibonacci number"
       (fact "given 0"
             (fibonacciNumber 0) => 0)
       (fact "given 1"
             (fibonacciNumber 1) => 1)
       (fact "given 2"
             (fibonacciNumber 2) => 2)
       (fact "given 3"
             (fibonacciNumber 3) => 3)
       (fact "given 4"
             (fibonacciNumber 4) => 5)
       (fact "given 5"
             (fibonacciNumber 5) => 8)
       (fact "given 6"
             (fibonacciNumber 6) => 13)
       (fact "given 10"
             (fibonacciNumber 10) => 89))

(facts "swap Case"
       (fact "given empty"
             (alternate-case "") => "")
       (fact "given char"
             (alternate-case "a") => "A"
             (alternate-case "A") => "a"
             (alternate-case "C") => "c")
       (fact "given string"
             (alternate-case "ab") => "AB"
             (alternate-case "aB") => "Ab"
             (alternate-case "aBcDefG") => "AbCdEFg"))

(fact "mexican wave"
      (wave "") => []
      (wave "a") => ["A"]
      (wave " a ") => [" A "]
      (wave "ab") => ["Ab" "aB"])

(facts "max-ball"
       (fact "no speed"
             (max-ball 0) => 0)
       (fact "examples"
             (max-ball 15) => 4
             (provided
              (ball-height 3 anything) => 0.808
              (ball-height 4 anything) => 0.881 ;; (partial > ..before..)
              (ball-height 5 anything) => -1
              (ball-height anything anything) => 0.1)))

(facts (fact "zero ms"
             (ball-height 0 15) => 0.0)
       (fact "after more ms"
             (ball-height 1 15) => (roughly 0.367 0.001)
             (provided
              (km-h=>m-s 15) => 4.166)
             (ball-height 2 15) => (roughly 0.637 0.001)
             (provided
              (km-h=>m-s 15) => 4.166)))

(facts "km-h=>m-s"
       (fact "zero"
             (km-h=>m-s 0) => 0.0)
       (fact "more"
             (km-h=>m-s 9) => 2.5))
(facts
 (fact "zero numbers"
       (nb-dig 0 0) => 1
       (nb-dig 0 1) => 0)
 (fact "one number"
       (nb-dig 1 0) => 1
       (nb-dig 1 1) => 1
       (nb-dig 1 2) => 0)
 (fact "squared number single digit"
       (nb-dig 2 0) => 1
       (nb-dig 2 1) => 1
       (nb-dig 2 2) => 0
       (nb-dig 2 4) => 1
       (nb-dig 3 3) => 0
       (nb-dig 3 0) => 1
       (nb-dig 3 1) => 1
       (nb-dig 3 4) => 1
       (nb-dig 3 9) => 1))

(fact "empty"
      (scale "" 0 0) => "")


test/clojure_katas/prime_test.clj
(ns clojure-katas.prime-test
  (:require [midje.sweet :refer [=> facts fact]]
            [clojure-katas.core :refer :all]))

(facts "test prime"
       (fact "zero"
             (primeFactors 0) => 0))


src/clojure_katas/ping_pong.clj
(ns clojure-katas.ping-pong
  (:import (javax.swing JFrame)
           (java.awt Color Font Dimension GraphicsEnvironment Toolkit)
           (java.awt.event KeyListener)))

; The pong in Clojure
;
; by Cesar Canassa and Julio Nobrega
;
; 2011-02-26
;
; References:
; http://gpwiki.org/index.php/Java:Tutorials:Double_Buffering
; http://zetcode.com/tutorials/javagamestutorial/
; http://jng.imagine27.com/articles/2009-09-12-122605_pong_in_clojure.html
; http://www.bestinclass.dk/index.clj/2010/10/taking-uncle-bob-to-school.html
; http://www.youtube.com/watch?v=nzHDQvzUXL0

; This atom stores if the racquet is going up (1) down (-1) or is stopped (0)
(def racquet-left-state (atom {:up false :down false}))
(def racquet-right-state (atom {:up false :down false}))

(def *paused* 0)
(def *running* 1)
(def game-state (atom *paused*))

; On scoring, ball will be thrown from an arc between -0.2 and 0.2
(defn rand-ball-arc
  []
  (/ (- (rand) 0.8) 5)
  )

;;;;;;;;;;;;;;;;; Colision checking ;;;;;;;;;;;;;;;;;
(defn colision-top?
  [game]
  (< ((game :ball) :y) (+ (game :bleacher-height) (game :lane-size))))

(defn colision-bottom?
  [game]
  (> ((game :ball) :y) (- (game :window-height) (game :ball-size))))

(defn colision-right?
  [game]
  (> ((game :ball) :x) (- (game :window-width) (game :ball-size))))

(defn colision-left?
  [game]
  (< ((game :ball) :x) 0))

(defn colision-racquet-left?
  [game]
  (let [ball (game :ball)
        racquet (game :racquet-left-pos)
        top (- racquet (game :racquet-middle-height))
        bottom (+ racquet (game :racquet-middle-height))]
    (and (< (ball :x) (+ (game :racquet-distance) (game :racquet-width)))
         (> (ball :y) top)
         (< (ball :y) bottom))))

(defn colision-racquet-right?
  [game]
  (let [ball (game :ball)
        racquet (game :racquet-right-pos)
        top (- racquet (game :racquet-middle-height))
        bottom (+ racquet (game :racquet-middle-height))]
    (and (> (ball :x) (- (game :window-width) (game :ball-size) (game :racquet-width) (game :racquet-distance)))
         (> (ball :y) top)
         (< (ball :y) bottom))))

;;;;;;;;;;;;;;;;; Collision actions ;;;;;;;;;;;;;;;;;
(defn collided-right
  [game]
  (let [ball (game :ball)]
    (merge game {:ball (merge ball {:x (* (/ (game :window-width) 4) 3)
                                    :y (+ (game :bleacher-height) (rand-int (game :window-height)))
                                    :sx (- (ball :ball-starting-sx))
                                    :sy (rand-ball-arc)
                                    })
                 :player-left-score (inc (game :player-left-score))
                 })))

(defn collided-racquet-right
  [game]
  (let [ball (game :ball)
        racquet (game :racquet-right-pos)
        hit (/ (- (ball :y) racquet) (game :racquet-middle-height))]
    (merge game {:ball (merge ball {:x (- (game :window-width) (game :ball-size) (game :racquet-width) (game :racquet-distance))
                                    :sx (* -1 (Math/cos hit) (game :speed))
                                    :sy (* (Math/sin hit) (game :speed))})
                 :speed (+ (game :speed) (game :increment))})))

(defn collided-left
  [game]
  (let [ball (game :ball)]
    (merge game {:ball (merge ball {:x (/ (game :window-width) 4)
                                    :y (+ (game :bleacher-height) (rand-int (game :window-height)))
                                    :sx (ball :ball-starting-sx)
                                    :sy (rand-ball-arc)
                                    })
                 :player-right-score (inc (game :player-right-score))
                 })))

(defn collided-racquet-left
  [game]
  (let [ball (game :ball)
        racquet (game :racquet-left-pos)
        hit (/ (- (ball :y) racquet) (game :racquet-middle-height))]
    (merge game {:ball (merge ball {:x (+ (game :racquet-distance) (game :racquet-width))
                                    :sx (* (Math/cos hit) (game :speed))
                                    :sy (* (Math/sin hit) (game :speed))})
                 :speed (+ (game :speed) (game :increment))})))

;;;;;;;;;;;;;;;;; Object updates ;;;;;;;;;;;;;;;;;
(defn update-ball
  [game step]
  (let [ball (game :ball)]
    ; The cond form is usually a bad ideia. There should a better way to do this.
    (cond
      ; This requires some serious DRY
      (colision-racquet-left? game) (collided-racquet-left game)
      (colision-racquet-right? game) (collided-racquet-right game)
      (colision-top? game) (merge game {:ball (merge ball {:y (+ (game :bleacher-height) (game :lane-size))
                                                           :sy (* -1 (ball :sy))})})
      (colision-bottom? game) (merge game {:ball (merge ball {:y (- (game :window-height) (game :ball-size))
                                                              :sy (* -1 (ball :sy))})})
      ; Apply the physics
      :else (merge game {:ball (merge ball {:x (+ (ball :x) (* step (ball :sx)))
                                            :y (+ (ball :y) (* step (ball :sy)))})}))))

(defn update-racquet
  [game position state step]
  (let [top (+ (game :bleacher-height)
               (game :racquet-middle-height))
        bottom (- (game :window-height) (game :racquet-middle-height))]
    (cond
      ; Collisions
      (< position top) top
      (> position bottom) bottom
      ; Position updates
      (and (= (state :up) true) (= (state :down) true)) position
      (= (state :up) true) (- position (* step (game :racquet-speed)))
      (= (state :down) true) (+ position (* step (game :racquet-speed)))
      :else position)))

(defn update-game
  [game step]
  (let [game (update-ball game step)
        racquet-left (update-racquet game (game :racquet-left-pos) @racquet-left-state step)
        racquet-right (update-racquet game (game :racquet-right-pos) @racquet-right-state step)]

    (cond
      (colision-right? game) (collided-right game)
      (colision-left? game) (collided-left game)
      :else
      (merge game {:ball (game :ball)
                   :racquet-left-pos racquet-left
                   :racquet-right-pos racquet-right}))))

;;;;;;;;;;;;;;;;; Draw, Keypress, Main loop ;;;;;;;;;;;;;;;;;
(defn drawn
  [frame game fps]
  (let [buffer (.getBufferStrategy frame)
        graphics (.getDrawGraphics buffer)
        ball (game :ball)
        racquet-left-position (game :racquet-left-pos)
        racquet-right-position (game :racquet-right-pos)]

    (doto graphics
      ; Clears the screen
      (.setColor Color/BLACK)
      (.fillRect 0 0 (game :window-width) (game :window-height))

      (.setColor Color/WHITE)

      ; Draw the ball
      (.fillOval (ball :x) (ball :y) (game :ball-size) (game :ball-size))

      ; Draw the court top lane
      (.fillRect 0 (- (game :bleacher-height) (game :lane-size)) (game :court-width) (game :lane-size))

      ; Draw the court division lane
      (.fillRect (- (/ (game :court-width) 2) (game :lane-size))
                 (game :bleacher-height)
                 (game :lane-size)
                 (game :court-height))

      ; Draw both racquets
      (.fillRect (game :racquet-distance)
                 (- (game :racquet-left-pos) (game :racquet-middle-height))
                 (game :racquet-width) (game :racquet-height))
      (.fillRect (- (game :window-width) (+ (game :racquet-width) (game :racquet-distance)))
                 (- (game :racquet-right-pos) (game :racquet-middle-height))
                 (game :racquet-width)
                 (game :racquet-height))

      ; Draw both scores
      (.setFont (new Font "Courier" (. Font BOLD) 50))
      (.drawString (str (game :player-left-score)) (/ (game :window-width) 4) (game :score-height))
      (.drawString (str (game :player-right-score)) (* (/ (game :window-width) 4) 3) (game :score-height))

      ; Draw FPS counter
      (.setFont (new Font "Courier" (. Font PLAIN) 20))
      (.drawString (str "FPS: " fps) 770 20))

    (when (= @game-state *paused*)
      (let [font (new Font "Courier" (. Font BOLD) 100)
            metrics (.getFontMetrics graphics font)
            paused "GAME PAUSED"
            paused-width (.stringWidth metrics paused)
            paused-x (/ (- (game :window-width) paused-width) 2)]
        (.setFont graphics font)
        (.setColor graphics Color/RED)
        (.drawString graphics "GAME PAUSED" paused-x (/ (game :window-height) 2))
        (.setFont graphics (new Font "Courier" (. Font BOLD) 50))
        (.drawString graphics "p: Start/Pause" paused-x (+ (/ (game :window-height) 2) 60))
        (.drawString graphics "a z: Move left racquet" paused-x (+ (/ (game :window-height) 2) 110))
        (.drawString graphics "j m: Move right racquet" paused-x (+ (/ (game :window-height) 2) 160))
        (.drawString graphics "q: Quit" paused-x (+ (/ (game :window-height) 2) 210))
        ))
    ; It is best to dispose() a Graphics object when done with it.
    (.dispose graphics)

    ; Shows the contents of the backbuffer on the screen.
    (.show buffer)))

(defn swap-key
  [atom struct]
  (swap! atom merge @atom struct))

(defn handle-keypress
  [e]
  (case e
    ; Exits when 'q' is pressed
    \q (System/exit 0)

    \p (reset! game-state (if (= @game-state *paused*) *running* *paused*))

    ; Pressing 'a' or 'z' updates the left racquet state
    \a (swap-key racquet-left-state {:up true})
    \z (swap-key racquet-left-state {:down true})

    ; Pressing 'j' or 'm' updates the right racquet state
    \j (swap-key racquet-right-state {:up true})
    \m (swap-key racquet-right-state {:down true})
    nil))

(defn handle-keyrelease
  [e]
  ; Releasing the keys stops the racquet
  (case e
    \a (swap-key racquet-left-state {:up false})
    \z (swap-key racquet-left-state {:down false})
    \j (swap-key racquet-right-state {:up false})
    \m (swap-key racquet-right-state {:down false})
    nil))

(defn new-game
  [width height]
  (let [bleacher-height 200
        court-height (- height bleacher-height)
        racquet-height (/ court-height 5)
        ball-starting-sx 0.3]
    {:ball {:x (/ width 2)
            :y (+ (/ court-height 2) bleacher-height)
            :ball-starting-sx ball-starting-sx
            :sx ball-starting-sx
            :sy (rand-ball-arc)}
     :speed 0.5
     :increment 0.05
     :player-left-score 0
     :player-right-score 0
     :racquet-left-pos 400
     :racquet-right-pos 400
     :state *paused*

     :window-width width
     :window-height height
     :bleacher-height bleacher-height

     :lane-size 5
     :ball-size 10
     :score-height (/ bleacher-height 2)

     :court-width width
     :court-height court-height

     :racquet-distance 10 ; How far from the court side walls
     :racquet-speed 0.6 ; How fast the racquet moves
     :racquet-width 10
     :racquet-height racquet-height
     :racquet-middle-height (/ racquet-height 2)}))

(defn -main
  []
  (let [frame (new JFrame "Clojure Pong")
        start-time (System/currentTimeMillis)
        toolkit (. Toolkit getDefaultToolkit)
        ge (GraphicsEnvironment/getLocalGraphicsEnvironment)
        gd (. ge getDefaultScreenDevice)]
    (doto frame
      (.setDefaultCloseOperation JFrame/EXIT_ON_CLOSE)
      (.setUndecorated true)
      ;(.setIgnoreRepaint true) ; This is supposed to give us extra FPS, but I see to diference
      (.setResizable false))

    (.setFullScreenWindow gd frame)

    (doto frame
      (.setVisible true)
      (.createBufferStrategy 2)

      (.addKeyListener
        (proxy [KeyListener] []
          (keyPressed [e]
            (handle-keypress (.getKeyChar e)))
          (keyReleased [e]
            (handle-keyrelease (.getKeyChar e)))
          (keyTyped [e])))

      ; Makes sure everything inside the frame fits
      (.validate)

      (.show))

    (loop [time start-time
           old-time start-time
           game (new-game (.. toolkit getScreenSize width) (.. toolkit getScreenSize height))
           fps 0
           frame-counter 0
           one-second 0]
      (let [step (- time old-time)
            new-fps? (>= one-second 1000)]

        (drawn frame game fps)

        (recur (System/currentTimeMillis)
               time
               (if (= @game-state *running*) (update-game game step) game)
               (if new-fps? frame-counter fps)
               (if new-fps? 0 (inc frame-counter))
               (if new-fps? 0 (+ one-second step)))))))


src/clojure_katas/core.clj
(ns clojure-katas.core
  (:require [clojure.string :as string]))

(defn divides? [input divider]
  (= (mod input divider) 0))

(defn fizzBuzz [input]
  (cond
    (zero? input) 0
    (divides? input 15) "fizzbuzz"
    (divides? input 3) "fizz"
    (divides? input 5) "buzz"
    :else input))

(defn primeFactors [input]
  0)

(defn factorial [input]
  (cond (zero? input) 1
        :else
        (* input (factorial (- input 1)))))

(defn spop [])
  

(defn add100 [number]
  (+ number 100))

(defn dec-maker
  [toDec]
  #(- % toDec))

(defn mapset
  [operation & input]
  (hash-set (map operation input)))

(def asym-hobbit-body-parts [{:name "head" :size 3}
                             {:name "left-eye" :size 1}
                             {:name "left-ear" :size 1}
                             {:name "mouth" :size 1}
                             {:name "nose" :size 1}
                             {:name "neck" :size 2}
                             {:name "left-shoulder" :size 3}
                             {:name "left-upper-arm" :size 3}
                             {:name "chest" :size 10}
                             {:name "back" :size 10}
                             {:name "left-forearm" :size 3}
                             {:name "abdomen" :size 6}
                             {:name "left-kidney" :size 1}
                             {:name "left-hand" :size 2}
                             {:name "left-knee" :size 2}
                             {:name "left-thigh" :size 4}
                             {:name "left-lower-leg" :size 3}
                             {:name "left-achilles" :size 1}
                             {:name "left-foot" :size 2}])

(defn matching-part
  [part]
  {:name (clojure.string/replace (:name part) #"^left-" "right-")
   :size (:size part)})

(defn symmetrize-body-parts
  "Expects a seq of maps that have a :name and :size"
  [asym-body-parts]
  (loop [remaining-asym-parts asym-body-parts
         final-body-parts []]
    (if (empty? remaining-asym-parts)
      final-body-parts
      (let [[part & remaining] remaining-asym-parts]
        (recur remaining
               (into final-body-parts
                     (set [part (matching-part part)])))))))

(defn better-symmetrize-body-parts
  "Expects a seq of maps that have a :name and :size"
  [asym-body-parts]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set [part (matching-part part)])))
          []
          asym-body-parts))

(defn matching-part-alien
  [part]
  (hash-set
    {:name (:name part)
     :size (:size part)}
    {:name (clojure.string/replace (:name part) #"^left-" "right-")
     :size (:size part)}
    {:name (clojure.string/replace (:name part) #"^left-" "down-")
     :size (:size part)}
    {:name (clojure.string/replace (:name part) #"^left-" "up-")
     :size (:size part)}
    {:name (clojure.string/replace (:name part) #"^left-" "middle-")
     :size (:size part)}))
(defn multiples3And5 [number]
   (reduce + (filter #(or (divides? % 3) (divides? % 5)) (range 1 number))))

(defn fibonacciNumber [number]
  (if (or (= number 1) (= number 2) (= number 0))
    number
    (+ (fibonacciNumber (- number 1)) (fibonacciNumber (- number 2)))))

(def fib-seq
  ((fn rfib [a b]
     (lazy-seq (cons a (rfib b (+ a b)))))
   0 1))

(defn fib [start range]
  "Creates a vector of fibonnaci numbers"
  (if (<= range 0)
    start
    (recur (let[subvector (subvec start (- (count start) 2))
                x (nth subvector 0)
                y (nth subvector 1)
                z (+ x y)]
             (conj start z))
           (- range 1))))

(defn fib-step [[a b]]
  [b (+ a b)])

(defn fib-seq []
  (map first (iterate fib-step [0 1])))

(defn sumOfEvenFibonacciNumbers [max]
  (reduce + (filter even? (take max (fib-seq)))))

(def vampire-database
  {0 {:makes-blood-puns? false, :has-pulse? true :name "McFishwich"}
   1 {:makes-blood-puns? false, :has-pulse? true :name "McMackson"}
   2 {:makes-blood-puns? true, :has-pulse? false :name "Damon Salvatore"}
   3 {:makes-blood-puns? true, :has-pulse? true :name "Mickey Mouse"}})

(defn vampire-related-details
  [social-security-number]
  (Thread/sleep 1000)
  (get vampire-database social-security-number))

(defn vampire?
  [record]
  (and (:makes-blood-puns? record)
       (not (:has-pulse? record))
       record))

(defn identify-vampire
  [social-security-numbers]
  (first (filter vampire?
                 (map vampire-related-details social-security-numbers))))

(defn- alternate-char-case
  [s]
  (if (= (str s) (string/upper-case s))
    (string/lower-case s)
    (string/upper-case s)))
    

(defn alternate-case [s]
  (apply str (map alternate-char-case (seq s))))

(defn- update-string-at [input index]
  (apply str (update (vec input) index #(Character/toUpperCase %))))

(defn- get-non-space-char-indexes [input]
  (filter #(Character/isLetter (nth input %)) (range (count input))))

(defn wave [input]
    (map #(update-string-at input %)
        (get-non-space-char-indexes input)))
         ;; (let [indexes (get-non-space-char-indexes input)]
         ;;   (if (seq indexes)
         ;;     indexes
         ;;     '(0))))))

(def ^:const conversion-multiplier (/ 1000 3600))
(def ^:const gravitational-constant 9.81)

(defn km-h=>m-s [speed]
  (float (* speed conversion-multiplier)))

(defn ball-height [time speed-in-km-h]
  (let [t (* 0.1 time)]
    (- (* (km-h=>m-s speed-in-km-h) t)
       (* 0.5 t t gravitational-constant))))
  
(defn max-ball [speed-in-km-h]
  (let [heights
        (take-while
               #(<= 0 %)
               (map #(ball-height % speed-in-km-h) (range)))]
      (.indexOf heights (apply max heights))))  

(defn squares-till
  [max-val]
  (map #(* % %) (range (inc max-val))))

(defn nb-dig [max-val digit]
  (count (filter #{(.charAt (str digit) 0)} (apply str (squares-till max-val)))))

(defn scale [strng hor ver]
  "")
